File: Strategy/CSharp/Strategy/Ducks/Duck.cs
using System;
using Strategy;
using Strategy.Behaivours;

namespace Strategy.Ducks
{
	public abstract class Duck
	{
		IFlyBehaviour flybehaviour;
		IQuackBehaviour quackbehavior;
		
		public Duck ()
		{
		}
		
		public void setFlyBehaivour(IFlyBehaviour behaviour)
		{
			flybehaviour = behaviour;
		}
		
		public void performFly()
		{
			flybehaviour.fly();
		}
	}
}

File: Strategy/PHP/Strategy/Ducks/Duck.php
<?php
 /**
 * Duck.php
 * 
 * An abstract Duck
 * This class can be seen as the context of the strategy
 * @author Indika Piyasena
 * @version 1.0
 */
 
abstract class Duck {

	protected $flyBehaviour;

	/**
	 * Set an flyBehaviour
	 * @param  $flyBehaviour IFlyBehaviour
	 * @return bool TRUE, if the provided parser implements IFlyBehaviour
	 */
	public function setFlyBehaviour($flyBehaviour)
	{
		if (is_a($flyBehaviour, 'IFlyBehaviour'))
		{
			$this->flyBehaviour = $flyBehaviour;
		    return TRUE;
		}
		else
		{
			//log_message('error', ': the provided flyBehaviour does not implement IFlyBehaviour');
		    return FALSE;
		}
	}

	public function performFly()
	{
		$this->flyBehaviour->fly();
	}


}
